src/workerd/api/sockets.h

```diff
@@ -14,7 +15,7 @@ class PipelinedAsyncIoStream final : public kj::AsyncIoStream, public kj::Refcou
   // A stream that is backed by a promise for an AsyncIoStream. All operations on this stream
   // are deferred until the `inner` promise completes.
 public:
-  explicit PipelinedAsyncIoStream(kj::Promise<kj::Own<kj::AsyncIoStream>> inner);
+  explicit PipelinedAsyncIoStream(kj::Promise<kj::Own<RevocableIoStream>> inner);
```

Comment:
Is this still needed? There is `kj::newPromisedStream(...)` in KJ, and the stream that you get back from the connect includes proper guards on writes and reads. I think it might be possible to replace this entirely with `kj::newPromisedStream`

Reply:
Let's keep it for now, I'll need it for one of my follow up PRs and I'll rename it then (it'll be needed to get notified when the stream reads 0 bytes so I can close the socket implicitly)

Reply:
Ok, that said, I don't think you actually need to change the `kj::AsyncIoStream` here to `Revocable`... unless I missed something?

Reply:
Yeah, I was puzzled by this myself, clang just wouldn't accept `kj::AsyncIoStream` here for some reason. Though maybe it was some other piece of code causing the issue, I will try to revert that and see if it still happens.

Reply:
Yeah, still getting the same error:

```
bazel-out/k8-dbg/bin/external/capnp-cpp/src/kj/_virtual_includes/kj/kj/refcount.h:95:41: error: no matching constructor for initialization of 'workerd::api::PipelinedAsyncIoStream'
  return Refcounted::addRefInternal(new T(kj::fwd<Params>(params)...));
                                        ^ ~~~~~~~~~~~~~~~~~~~~~~~
external/workerd/src/workerd/api/sockets.c++:62:21: note: in instantiation of function template specialization 'kj::refcounted<workerd::api::PipelinedAsyncIoStream, kj::Promise<kj::Own<workerd::api::RevocableIoStream>>>' requested here
  auto stream = kj::refcounted<PipelinedAsyncIoStream>(kj::mv(connectionPromise));
                    ^
external/workerd/src/workerd/api/sockets.h:14:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'kj::Promise<kj::Own<workerd::api::RevocableIoStream>>' to 'const workerd::api::PipelinedAsyncIoStream' for 1st argument
class PipelinedAsyncIoStream final : public kj::AsyncIoStream, public kj::Refcounted {
      ^
external/workerd/src/workerd/api/sockets.c++:88:25: note: candidate constructor not viable: no known conversion from 'Promise<Own<workerd::api::RevocableIoStream, [...]>>' to 'Promise<Own<kj::AsyncIoStream, [...]>>' for 1st argument
PipelinedAsyncIoStream::PipelinedAsyncIoStream(kj::Promise<kj::Own<kj::AsyncIoStream>> inner) :
                        ^
1 warning and 1 error generated.
Target @workerd//src/workerd/server:workerd failed to build
```

Maybe I'm missing something but:

```
candidate constructor not viable: no known conversion from 'Promise<Own<workerd::api::RevocableIoStream, [...]>>' to 'Promise<Own<kj::AsyncIoStream, [...]>>' for 1st argument
```

Tells me I cannot pass `RevocableIoStream` where an `kj::AsyncIoStream` is expected (even though RevocableIoStream extends kj::AsyncIoStream, which puzzles me)

Reply:
It's a limitation of the template definitions, I think. if you do something like:

```cpp
kj::Own<kj::AsyncIoStream> io = getRevocableSomehow();
```

And pass `io` on, it should work, I believe. But this is fine for now.